prover/src/logup_gkr/mod.rs

```diff
@@ -113,46 +115,105 @@ impl<E: FieldElement> EvaluatedCircuit<E> {
         let num_fractions = evaluator.get_num_fractions();
         let periodic_values = evaluator.build_periodic_values();
 
-        let mut input_layer_wires =
-            Vec::with_capacity(main_trace.main_segment().num_rows() * num_fractions);
-        let mut main_frame = EvaluationFrame::new(main_trace.main_segment().num_cols());
-
-        let mut query = vec![E::BaseField::ZERO; evaluator.get_oracles().len()];
-        let mut periodic_values_row = vec![E::BaseField::ZERO; periodic_values.num_columns()];
-        let mut numerators = vec![E::ZERO; num_fractions];
-        let mut denominators = vec![E::ZERO; num_fractions];
-        for i in 0..main_trace.main_segment().num_rows() {
-            let wires_from_trace_row = {
-                main_trace.read_main_frame(i, &mut main_frame);
-                periodic_values.fill_periodic_values_at(i, &mut periodic_values_row);
-                evaluator.build_query(&main_frame, &mut query);
-
-                evaluator.evaluate_query(
-                    &query,
-                    &periodic_values_row,
-                    log_up_randomness,
-                    &mut numerators,
-                    &mut denominators,
-                );
-                let input_gates_values: Vec<CircuitWire<E>> = numerators
-                    .iter()
-                    .zip(denominators.iter())
-                    .map(|(numerator, denominator)| CircuitWire::new(*numerator, *denominator))
-                    .collect();
-                input_gates_values
+        #[cfg(feature = "concurrent")]
+        let input_layer_wires = {
+            let mut input_layer_wires = unsafe {
+                utils::uninit_vector(main_trace.main_segment().num_rows() * num_fractions)
```

Comment:
I wonder if we could use the `batch_iter_mut!` macro here. Maybe something like this:
```rust
batch_iter_mut!(&mut input_layer_wires, 1024, |batch: &mut [E], batch_offset: usize| {
    // run a loop here similar to the one on lines 185 - 207 but taking the offset into account
}
```
If this works, it may speed things up a bit and should also allow us to use the same code for concurrent and serial versions.

Reply:
Implemented!
The performance is the same as far as I am seeing, but the removing code duplication makes this proposal better.